### ISSUE TITLE
 Data types are not strict

### ISSUE THREAD

**Issue by ofttryaj** [symfony/symfony#19867]
Symfony\Component\HttpKernel\Kernel

``` php
// this container's type is ContainerInterface
protected function getContainerLoader(ContainerInterface $container)
    {
        $locator = new FileLocator($this);
        $resolver = new LoaderResolver(array(
            new XmlFileLoader($container, $locator),
            new YamlFileLoader($container, $locator),
            new IniFileLoader($container, $locator),
            new PhpFileLoader($container, $locator),
            new DirectoryLoader($container, $locator),
            new ClosureLoader($container),
        ));

        return new DelegatingLoader($resolver);
    }
```

Symfony\Component\DependencyInjection\Loader\FileLoader

``` php
// this container's type is ContainerBuilder, but the above, 
// this construct function is called, params type is ContainerInterface

public function __construct(ContainerBuilder $container, FileLocatorInterface $locator)
    {
        $this->container = $container;

        parent::__construct($locator);
    }
```

**Comment by HeahDude** [comment]
I'm not sure to understand your issue here.
It is the responsibility of a method to type hint an interface or a specific implementation. And we should not change one signature because of another systematically.

**Comment by ofttryaj** [comment]
@HeahDude I'm very sorry. My English is very poor, i means Fileloader construct function， container parameter should be the type ContainerInterface.^_^

**Comment by ofttryaj** [comment]
![fileloader](https://cloud.githubusercontent.com/assets/15887451/18297004/2bb975f4-74e0-11e6-8a3f-160b36a5cf51.png)

**Comment by ro0NL** [comment]
Type hint of `getContainerLoader` should be `ContainerBuilder $container`.

Loaders expect a builder by design, and therefor this method as well. Which also happens: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Kernel.php#L566

I agree this need to be looked at.

**Comment by javiereguiluz** [comment]
Closing as fixed by #19960.
